Update docs and hooks worker client for go hooks

ddelnano opened the pull request:
This PR depends on [this](https://github.com/snikch/goodman/pull/5).  Once that is merged, the go hooks should be considered compliant with the hooks for other languages: php, ruby, python, perl, etc.

I tested the changes with this [repo](https://github.com/ddelnano/dredd-hooks-go-example).  I could not install dredd from a github repo since npm does not run prepublish and the necessary commands to do so I made these commits on my fork and made sure the examples worked.

<img width="782" alt="screen shot 2016-05-31 at 10 58 52 pm" src="https://cloud.githubusercontent.com/assets/5855593/15696950/6532fa62-2783-11e6-975c-5d6b43214794.png">

ddelnano commented:
@honzajavorek The PR this depended on is finally [merged](https://github.com/snikch/goodman/pull/5)

honzajavorek commented:
@ddelnano Great! I asked our Go lover @pksunkara to look at it once he has a free moment, so expect him to land & comment here soon.

ddelnano commented:
@honzajavorek I am having problems with the Go hooks dealing with requests that have an empty body.  Like the ones in the hooks template.  If I add a request body all the aruba/cucumber tests pass.  If I don't I get this dredd error.
<img width="1044" alt="screen shot 2016-06-03 at 9 00 04 am" src="https://cloud.githubusercontent.com/assets/5855593/15779313/97827a00-2969-11e6-8278-75c7d99183e2.png">

Seems to me to be kind of like the error in #499 where it shouldn't allow whatever value that is to be passed to Buffer.byteLength, but I think there is some other issue where maybe the go hooks send back bad data or something that causes dredd to have these errors.

ddelnano commented:
@honzajavorek so the CI build for the go hooks pass if I add this to every .feature file

https://github.com/snikch/goodman/blob/master/features/execution_order.feature#L19-L20

ddelnano commented:
@honzajavorek deleted my previous comments, cause I thought this was on an issue and not this PR.  Sounds good to me 👍

pksunkara commented on docs/hooks-go.md:
```diff
@@ -10,21 +10,49 @@ Go hooks are using [Dredd's hooks handler socket interface](hooks-new-language.m
 ## Usage
 
-Using Dredd with Go is slightly different to other languages, as a binary needs to be compiled for execution. Instead of providing a variety of hookfiles to run, your own binary is built using the Go hooks library `goodman` to run your specific hooks. The path to this binary is passed to dredd via the `--language` flag. You may also need to supply a `--hookfiles` flag to get the tests to run, but ...
+Using Dredd with Go is slightly different to other languages, as a binary needs to be compiled for execution. The --hookfiles flags will be compiled hook binaries.  See below for an example hooks.go file to get an idea of what this would look like.
```
> hookfiles flags should point to compiled hook binaries.

> get an idea of what the source file behind the go binary would look like.

ddelnano replied:
@pksunkara can you be more specific?  Is this just not clear?

pksunkara replied:
I just made the sentences a bit more clear. My comment has the updated sentences which you need to copy.

ddelnano replied:
@pksunkara awesome thanks for clarifying.  I updated it.

pksunkara commented on docs/hooks-go.md:
````diff
@@ -10,21 +10,49 @@ Go hooks are using [Dredd's hooks handler socket interface](hooks-new-language.m
 
 ```
-$ dredd apiary.apib http://localhost:3000 --language ./my-dredd-hook-server --hookfiles *.go
+$ dredd apiary.apib http://localhost:3000 --server ./go-lang-web-server-to-test --language go --hookfiles ./hook-file-binary
````
@honzajavorek What is this server option used for? I thought localhost:3000 is the server.

pksunkara replied:
@ddelnano If you are talking about hook server, maybe rename it to `./golang-hook-server`

ddelnano replied:
@pksunkara the --server option is the command to get the server that will be under test started.  So the command assumes that the `./go-lang-web-server` is a go web service that will start a webserver that listens on localhost port 3000.

ddelnano replied:
@pksunkara That is not referring to the hook server.

pksunkara replied:
Ah, thanks. I didn't know that.

ddelnano replied:
@pksunkara the hooks server command gets set [here](https://github.com/apiaryio/dredd/pull/505/files#diff-fab7c8fd00434dbb9259f70a29405350R138).

pksunkara commented on docs/hooks-go.md:
```diff
@@ -130,21 +156,22 @@ package main
-    runner.Before("Machines > Machines collection > Post Machines", func(t *goodman.Transaction) {
-      t.Fail = "POST is broken"
+    h.Before("Machines > Machines collection > Post  Machines", func(t *trans.Transaction) {
+        t.Fail = "Post is broken"
```
`POST`

ddelnano replied:
@pksunkara updated.

pksunkara commented:
Just 2 minor comments regarding docs. Will merge it after they are fixed.

ddelnano commented:
@pksunkara @honzajavorek I updated the PR addressing the feedback **before** this get merged, however, I am thinking there might be something that might need addressed first.  Dredd crashes with the Go hooks if the API Blueprint transaction has an empty request body.  See the difference between the Go [tests](https://github.com/snikch/goodman/blob/master/features/execution_order.feature#L19-L20) and the template [tests](https://github.com/apiaryio/dredd-hooks-template/blob/master/features/execution_order.feature#L16-L20)

ddelnano commented:
Here is an example of the dredd crash, I am going to try and prepare a fix.

<img width="1052" alt="screen shot 2016-06-03 at 1 27 54 pm" src="https://cloud.githubusercontent.com/assets/5855593/15787219/0b510968-298f-11e6-86ba-ccb0e490a503.png">

ddelnano commented:
@honzajavorek not sure exactly what the deal is with the findings in my previous comments.  I wrote out a full example using a sample go app, goodman, and dredd compiled from this branch and blueprint files that have empty request bodies run fine.  You can see the example [here](https://github.com/snikch/goodman/pull/7).  Either of you could run the example if you wanted to following the instructions in the Wiki to verify my claims that it works as expected, however, you would need to update the path in the [Makefile](https://github.com/snikch/goodman/pull/7/files#diff-e8849b97cf69febe9876e8ab47d44460R10) since I needed to use dredd compiled from this branch (since it uses these changes).

Let me know what you think.

pksunkara commented:
Does the above crash happen with other language hook files or only golang? Depending on that, we need to fix it in a separate PR. I am merging this for now.

ddelnano commented:
@pksunkara when I developed the php hooks I never ran into this.  I also tried to reproduce it with a real example instead of the cucumber tests (where I was seeing it) and I was unable to do so.  I opened a PR for another crash that I found that indicates a hook server failure ( #511 ).  This would just provide better information to the hook server dev instead of seeing the error `cannot read property 'length' of undefined`.

ddelnano commented:
@pksunkara @honzajavorek when do the pre releases go out?  I would like to merge my example in the goodman repo but since npm's support for installing repos via commit hashes does not run any building (so the coffeescript would not be compiled) I can't merge until this change is in a new release.

honzajavorek commented:
@ddelnano I should release another one within a day.

ddelnano commented:
awesome thanks!
